### PR TITLE
Add fix permissions

### PR DESCRIPTION
**Pull request by rohank07**
# Description

**What your PR adds/fixes/removes**
-Add fix-permissions to pip/conda installs
-Removed .sh extension and modified references to scripts/remote-desktop/fix-permissions
Close https://github.com/StatCan/kubeflow-containers/issues/256
# Tests / Quality Checks

## Automated Testing/build and deployment
- [ ] Does the image pass CI successfully (build, pass vulnerability scan, and pass automated test suite)?
- [ ] If new features are added (new image, new binary, etc), have new automated tests been added to cover these?
- [ ] If new features are added that require in-cluster testing (e.g. a new feature that needs to interact with kubernetes), have you added the `auto-deploy` tag to the PR before pushing in order to build and push the image to ACR so you can test it in cluster as a custom image?

## JupyterLab extensions

- [ ] Are all extensions "enabled" (`jupyter labextension list` from inside the notebook)?

## VS Code tests

- [ ] Does VS Code open?
- [ ] Can you install extensions?

## Code review

- [ ] Have you added the `auto-deploy` tag to your PR before your most recent push to this repo?  This causes CI to build the image and push to our ACR, letting reviewers access the built image without having to create it themselves
- [ ] Have you chosen a reviewer, attached them as a reviewer to this PR, and messaged them with the SHA-pinned image name for the final image to test (e.g. `k8scc01covidacr.azurecr.io/machine-learning-notebook-cpu:746d058e2f37e004da5ca483d121bfb9e0545f2b`)?

### CODE REVIEW

**Comment by blairdrummond** [comment]
@rohank07  I have very sad news haha. This image is *larger* than before. Something has gone wrong.

Screenshot from 2021-06-02 16-15-35

I did a before-after diff and we've added 300 mb.

Can you [Squash the image](https://docs.docker.com/engine/reference/commandline/image_build/) into a single layer and get the size of that image to use as a baseline

From there we might need to debug by seeing which changes increased the image size and why

**Review by blairdrummond — CHANGES_REQUESTED** [review]
The image size has increased by 300mb, so something has gone wrong...

**Comment by blairdrummond** [comment]
See here on how to squash an image

https://github.com/StatCan/minio/pull/4/files

**Comment by rohank07** [comment]
> @rohank07 I have very sad news haha. This image is _larger_ than before. Something has gone wrong.
> 
> Screenshot from 2021-06-02 16-15-35
> 
> I did a before-after diff and we've added 300 mb.
> 
> Can you [Squash the image](https://docs.docker.com/engine/reference/commandline/image_build/) into a single layer and get the size of that image to use as a baseline
> 
> From there we might need to debug by seeing which changes increased the image size and why

After a --squash of the image, the size came out to be 10.4GB

**Comment by Ito-Matsuda** [comment]
@rohank07 when you finish and merge this in could you also update [kubeflow-manifest tags](https://github.com/StatCan/kubeflow-manifest/blob/master/kustomize/application/jupyter-web-app/configs/spawner_ui_config.yaml#L17) to the generated SHA for master? 

the stuff in 
```
spawnerFormDefaults:
  image:
    # The container Image for the user's Jupyter Notebook
    # If readonly, this value must be a member of the list below
    value: k8scc01covidacr.azurecr.io/jupyterlab-cpu:f25cad42
    # The list of available standard container Images
    options:
      - k8scc01covidacr.azurecr.io/rstudio:f25cad42
      - k8scc01covidacr.azurecr.io/jupyterlab-cpu:f25cad42
      - k8scc01covidacr.azurecr.io/jupyterlab-pytorch:f25cad42
      - k8scc01covidacr.azurecr.io/jupyterlab-tensorflow:f25cad42
      - k8scc01covidacr.azurecr.io/remote-desktop:75f84102
```

(The remote desktop one should also be whatever sha the jupyterlab ones since it's all in the same repo, it's different right now because we recently moved it. )

**Comment by ca-scribner** [comment]
# TLDR

Sorry for the slow response, I lost track of this one. 

I think the issue with `remote-desktop` getting larger is that the image is structured differently and permission-fixing would need to be different (sorry @rohank07, we didn't catch this and gave you bad instructions!).  For this issue, I suggest we skip `remote-desktop` but make a new issue in the backlog that says to fix this issue for `remote-desktop`

# Details

`remote-desktop` has python installed twice (once in `0_` and once in `6_`).  The `fix-permissions` call in the `0_` file is fixing permissions on a jovyan home dir and a conda install, neither of which exist in this image!  Conda isn't installed until `6_`, so we're fixing permissions on the wrong stuff. 
 Really, we'd want to fix-permissions in the `0_` file on the python that was installed immediately in the same command that installed it, eg:

```bash
RUN apt-get update --yes \    
        && apt-get install --yes python3-pip tini language-pack-fr \
        && rm -rf /var/lib/apt/lists/*
        && fix-permissions $WHEREVER_PYTHON_IS_INSTALLED
```

where I think the permission fixing should be done on `/usr/local/lib/python3.8/`.  And then any further pip3 calls would need the same permission fixing on that same directory

@blairdrummond, why does `remote-desktop` has several installs of python?  There's an apt install near the top of the image, then a conda install near the bottom.  Any reason why we're maintaining both?  I understand why we need python when exiting the `0_` file (the later `kubeflow` file needs python), but why install python rather than conda?  Feels like it just makes the python install more confusing (I think when users at the end type python, it is likely coming from conda and doesn't have all the things we installed).  

For continuity, would it make more sense to install conda in `0_` along with all packages we normally inherrit from docker-stacks, that way the environment is consistent with the other images?

# Trivia

I think technically there's actually **five** installs of python in this image.  I see python2.7, python3, python3.8, python3.9, and conda's python :p  Some of those might be linked but that's at least 4

File: docker-bits/3_Kubeflow.Dockerfile [hunk 1]
```diff
@@ -7,9 +7,10 @@ RUN pip3 --no-cache-dir install --quiet \
       'kubeflow-pytorchjob==0.1.3' \
       'kubeflow-tfjob==0.1.3' \
       'minio==5.0.10' \
-      'git+https://github.com/zachomedia/s3fs@8aa929f78666ff9e323cde7d9be9262db5a17985'
+      'git+https://github.com/zachomedia/s3fs@8aa929f78666ff9e323cde7d9be9262db5a17985' 
 
 # kfp-azure-databricks needs to be run after kfp
 RUN pip3 --no-cache-dir install --quiet \
```
Comment (ca-scribner) [inline docker-bits/3_Kubeflow.Dockerfile:13]: I think both of these `RUN` statements need `fix-permissions` because they could be running as root if whatever docker-bit happens before them leaves the user as root (this is the case for us - see some of the output `Dockerfile`s which have `user=root` at this stage).

This also points out that we should explicitly set the user to what we want it to be in these docker-bits.  I **think** this step can be installed as the jovyan user (which then means fix-permissions shouldn't be required...I think).  To do that you can add a user command at the top of the file.  Or, you can add a `user root` at the top, then add fix-permissions too

File: docker-bits/6_remote-desktop.Dockerfile [hunk 2]
```diff
@@ -21,11 +21,11 @@ ENV RESOURCES_PATH="/resources"
...
 RUN python3 -m pip install \
-    'git+git://github.com/Ito-Matsuda/jupyter-desktop-server#egg=jupyter-desktop-server'
+    'git+git://github.com/Ito-Matsuda/jupyter-desktop-server#egg=jupyter-desktop-server' 
 
-#Fix-permissions
-COPY remote-desktop/fix-permissions.sh /usr/bin/fix-permissions.sh
-RUN chmod u+x /usr/bin/fix-permissions.sh
+# #Fix-permissions
+# COPY remote-desktop/fix-permissions.sh /usr/bin/fix-permissions.sh
+# RUN chmod u+x /usr/bin/fix-permissions.sh
```
Comment (ca-scribner) [inline docker-bits/6_remote-desktop.Dockerfile:29]: We can remove this now?
```suggestion
```

File: docker-bits/0_Rocker.Dockerfile [hunk 3]
```diff
@@ -11,13 +11,18 @@ ENV NB_GID="100"
 USER root
 ENV PATH="/home/jovyan/.local/bin/:${PATH}"
 
+#Fix-permissions
+COPY remote-desktop/fix-permissions /usr/bin/fix-permissions
+RUN chmod u+x /usr/bin/fix-permissions
+
+
```
Comment (ca-scribner) [inline docker-bits/0_Rocker.Dockerfile:18]: Removing extra newline for formatting
```suggestion
```

File: docker-bits/6_remote-desktop.Dockerfile [hunk 4]
```diff
@@ -340,7 +340,7 @@ RUN pip3 install --force websockify==0.9.0 \
     && tar -xf /tmp/novnc.tar.gz -C /tmp/ \
     && mv /tmp/noVNC-${NO_VNC_VERSION} /opt/novnc \
     && rm /tmp/novnc.tar.gz \
-    && chown -R $NB_UID:$NB_GID /opt/novnc
+    && chown -R $NB_UID:$NB_GID /opt/novnc
```
Comment (ca-scribner) [inline docker-bits/6_remote-desktop.Dockerfile:343]: Remove trailing whitespace
```suggestion
    && chown -R $NB_UID:$NB_GID /opt/novnc
```

File: docker-bits/6_remote-desktop.Dockerfile [hunk 5]
```diff
@@ -21,11 +21,11 @@ ENV RESOURCES_PATH="/resources"
 RUN mkdir $RESOURCES_PATH
 
 RUN python3 -m pip install \
-    'git+git://github.com/Ito-Matsuda/jupyter-desktop-server#egg=jupyter-desktop-server'
+    'git+git://github.com/Ito-Matsuda/jupyter-desktop-server#egg=jupyter-desktop-server'
```
Comment (ca-scribner) [inline docker-bits/6_remote-desktop.Dockerfile:24]: Remove trailing whitespace
```suggestion
    'git+git://github.com/Ito-Matsuda/jupyter-desktop-server#egg=jupyter-desktop-server' 
```

File: docker-bits/5_DB-Drivers.Dockerfile [hunk 6]
```diff
@@ -2,4 +2,6 @@
 RUN apt-get update && \
     apt-get install -y alien unixodbc unixodbc-dev && \
     pip install --no-cache-dir --quiet 'pyodbc==4.0.30' && \
-    rm -rf /var/lib/apt/lists/*
+    rm -rf /var/lib/apt/lists/* && \
+    fix-permissions $CONDA_DIR && \
+    fix-permissions /home/$NB_USER
```
Comment (ca-scribner) [inline docker-bits/5_DB-Drivers.Dockerfile:7]: End files with empty line
```suggestion
    fix-permissions /home/$NB_USER

```

File: docker-bits/3_Kubeflow.Dockerfile [hunk 7]
```diff
@@ -7,9 +7,10 @@ RUN pip3 --no-cache-dir install --quiet \
...
       'fire==0.3.1' \
-      'git+https://github.com/kubeflow/pipelines@1d86111d8f152d3ed7506ea59cee1bfbc28abbf9#egg=kfp-azure-databricks&subdirectory=samples/contrib/azure-samples/kfp-azure-databricks'
+      'git+https://github.com/kubeflow/pipelines@1d86111d8f152d3ed7506ea59cee1bfbc28abbf9#egg=kfp-azure-databricks&subdirectory=samples/contrib/azure-samples/kfp-azure-databricks'
```
Comment (ca-scribner) [inline docker-bits/3_Kubeflow.Dockerfile:15]: Remove trailing whitespace (but the newline is good)
```suggestion
      'git+https://github.com/kubeflow/pipelines@1d86111d8f152d3ed7506ea59cee1bfbc28abbf9#egg=kfp-azure-databricks&subdirectory=samples/contrib/azure-samples/kfp-azure-databricks'
```

**Review by ca-scribner — CHANGES_REQUESTED** [review]
Some formatting suggestions plus missing fix-permissions in 3_kubeflow

**Review by ca-scribner — DISMISSED** [review]
lgtm!

**Review by ca-scribner — APPROVED** [review]
lgtm, so long as the build tests pass one last time (I made a tiny formatting change and they're rerunning now).

When you merge, make sure to squash and merge to avoid merging a bunch of commits to master.
